062-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,202 @@
+
+The reporters will also export JVM metrics similar to the ones exported by jmx_exporter. These are provided by the [Hotspot exports](io.prometheus.client.hotspot) from the Prometheus Java client.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it later to the reporters.
+
+This proposal will produce an implementation for each type of metrics reporter.
+- `KafkaPrometheusMetricsReporter` usable on brokers (for Kafka metrics) and on Kafka clients (including Connect and Streams)
+- `YammerPrometheusMetricsReporter` usable on brokers (for Yammer metrics)
+
+The Prometheus metrics registry is a singleton and the HTTP server will also be a singleton. This will allow applications to start multiple instances of the reporter (for example in applications containing multiple Kafka clients like Streams, Connect), and still collect all metrics via a single HTTP endpoint per JVM.
+
+The reporter for Kafka metrics will be usable outside of Strimzi by applications using Kafka clients. To do so applications will need to set the `metric.reporters` configuration to `KafkaPrometheusMetricsReporter` and set the reporter configurations accordingly for each Kafka client they instantiate.
+
+### Handling of non-numeric metrics
+
+Out of all the metrics emitted by Kafka brokers and clients, a few of them have non-numeric values. Prometheus only supports numeric values for metrics. When using jmx_exporter it is possible to write rules that move the non-numeric value to a label. For example we do it in [kafka-connect-metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/packaging/examples/metrics/kafka-connect-metrics.yaml#L103-L111) for the `status` metrics which has a string value.
+
+I propose to do this automatically in the reporters. For example a metric named: `kafka.connect:type=connector-task-metrics,connector="{connector}",task="{task}"<>status` with the value `running` will be converted into `kafka_connect_connector_task_metrics_status{connector="heartbeats",task="0",status="running"}` and its value will be set to `1.0`.
+
+## Affected/not affected projects
+
+The reporter will be a new project, `strimzi-metrics-reporter`. It will be used by Strimzi components but should also be usable without Strimzi, for example in client side applications.
+
+### strimzi-kafka-operator
+
+Reporters are not usable with ZooKeeper so Strimzi should keep support for jmx_exporter while it also supports ZooKeeper. As it's new code, reporters should first be released as a preview feature and disabled by default. Once KRaft becomes the preferred deployment option and the, we should make the reporter the recommended and default option to collect metrics over jmx_exporter.
```

Comment:
I think this should be split into three phases:
* First, the new exporter is added as an additional option to the Strimzi custom resource, but we keep supporting the JMX Exporter and use it in our examples as the main thing (with possibly some new examples of CRs and dashboards prepared as well, but not as the main thing)
* Next, we should keep supporting JMX Exporter, but switch to the new exporter as the _main_ example. Plus we possibly deprecate the JMX Exporter support.
* At some point much later, we might decide to drop the JMX Exporter support completely

The _preview disabled by default_ suggests some feature gate and some controller switch of what is used by changing the default. But that is not how it works here because it is in the custom resource API and you cannot switch it. Only the users can switch it.

Reply:
Thanks for the suggestion! That's a lot clearer, I'll use that.